assets/js/components/product-attribute-control/index.js

```diff
@@ -22,38 +22,84 @@ class ProductAttributeControl extends Component {
 		this.state = {
 			list: [],
 			loading: true,
+			attribute: 0,
+			termsLoading: true,
 		};
+
+		this.debouncedGetTerms = debounce( this.getTerms.bind( this ), 200 );
+		this.renderItem = this.renderItem.bind( this );
+		this.onSelectAttribute = this.onSelectAttribute.bind( this );
 	}
 
 	componentDidMount() {
-		const getTermsInAttribute = ( { id } ) => {
-			return apiFetch( {
-				path: addQueryArgs( `/wc-pb/v3/products/attributes/${ id }/terms`, {
-					per_page: -1,
-				} ),
-			} ).then( ( terms ) => terms.map( ( t ) => ( { ...t, parent: id } ) ) );
-		};
-
+		const { selected } = this.props;
 		apiFetch( {
 			path: addQueryArgs( '/wc-pb/v3/products/attributes', { per_page: -1 } ),
 		} )
-			.then( ( attributes ) => {
-				// Fetch the terms list for each attribute group, then flatten them into one list.
-				Promise.all( attributes.map( getTermsInAttribute ) ).then( ( results ) => {
-					const list = attributes.map( ( a ) => ( { ...a, parent: 0 } ) );
-					results.forEach( ( terms ) => {
-						list.push( ...terms );
-					} );
-					this.setState( { list, loading: false } );
+			.then( ( list ) => {
+				list = list.map( ( item ) => ( { ...item, parent: 0 } ) );
+				this.setState( ( { attribute } ) => {
+					if ( ! attribute && selected.length > 0 ) {
+						const item = find( list, { slug: selected[ 0 ].attr_slug } );
+						attribute = item ? item.id : 0;
+					}
+					return { list, attribute, loading: false };
 				} );
 			} )
 			.catch( () => {
 				this.setState( { list: [], loading: false } );
 			} );
 	}
 
+	componentDidUpdate( prevProps, prevState ) {
+		if ( prevState.attribute !== this.state.attribute ) {
+			this.debouncedGetTerms();
+		}
+	}
+
+	getTerms() {
+		const { attribute } = this.state;
+		if ( ! attribute ) {
+			return;
+		}
+
+		apiFetch( {
+			path: addQueryArgs( `/wc-pb/v3/products/attributes/${ attribute }/terms`, {
+				per_page: -1,
+			} ),
+		} )
+			.then( ( terms ) => {
+				terms = terms.map( ( term ) => ( { ...term, parent: attribute } ) );
+				this.setState( ( { list } ) => ( {
```

Comment:
This is working well as is, but I was also thinking that storing all attribute terms in an object might remove the delay when toggling back and forth between attributes instead of refetching on change.

```js
attributes: {
  color: [ 'red', 'blue' ],
  size: [ 'medium', 'large' ],
};
```
What do you think?

Reply:
That's a good idea, but I think I'll handle that in a follow up PR.